ocdata/core/bulk.py

```diff
@@ -0,0 +1,287 @@
+import os
+import pickle
+
+import ase
+import numpy as np
+import pymatgen
+from pymatgen.core.surface import (
+    SlabGenerator,
+    get_symmetrically_distinct_miller_indices,
+)
+from pymatgen.io.ase import AseAtomsAdaptor
+from pymatgen.symmetry.analyzer import SpacegroupAnalyzer
+
+from ocdata.configs.constants import COVALENT_MATERIALS_MPIDS, MAX_MILLER
+from ocdata.configs.paths import BULK_PKL_PATH, PRECOMPUTED_SLABS_DIR_PATH
+
+
+class Bulk:
+    """
+    Initializes a bulk object in one of 3 ways:
+    - Directly pass in an ase.Atoms object.
+    - Pass in index of bulk to select from bulk database.
+    - Randomly sample a bulk from bulk database.
+
+    Arguments
+    ---------
+    bulk_atoms: ase.Atoms
+        Bulk structure.
+    bulk_id_from_db: int
+        Index of bulk to select if not doing a random sample.
+    bulk_db_path: str
+        Path to bulk database.
+    precomputed_slabs_path: str
+        Path to folder of precomputed slabs.
+    mp_id: str
+        Materials Project ID of bulk to select if not doing a random sample.
+        [CURRENTLY NOT SUPPORTED.]
+    """
+
+    def __init__(
+        self,
+        bulk_atoms: ase.Atoms = None,
+        bulk_id_from_db: int = None,
+        # mp_id = None,
+        bulk_db_path: str = BULK_PKL_PATH,
+        precomputed_slabs_path: str = PRECOMPUTED_SLABS_DIR_PATH,
+    ):
+        self.bulk_id_from_db = bulk_id_from_db
+        self.bulk_db_path = bulk_db_path
+        self.precomputed_slabs_path = precomputed_slabs_path
+
+        if bulk_atoms is not None:
+            self.atoms = bulk_atoms.copy()
+            self.mpid = None
+        elif bulk_id_from_db is not None:
+            bulk_db = pickle.load(open(bulk_db_path, "rb"))
+            (
+                self.atoms,
+                self.mpid,
+                _,
+                _,
+            ) = bulk_db[bulk_id_from_db]
+        else:
+            bulk_db = pickle.load(open(bulk_db_path, "rb"))
+            bulk_id_from_db = np.random.randint(len(bulk_db))
+            (
+                self.atoms,
+                self.mpid,
+                _,
+                _,
+            ) = bulk_db[bulk_id_from_db]
+            self.bulk_id_from_db = bulk_id_from_db
+
+        # Comment(@abhshkdz): Do we need this?
+        self.n_elems = len(set(self.atoms.symbols))
+
+    def set_mpid(self, mpid: str):
+        self.mpid = mpid
+
+    def set_bulk_id_from_db(self, bulk_id_from_db: int):
+        self.bulk_id_from_db = bulk_id_from_db
+
+    def get_slabs(self):
+        """
+        Returns a list of possible slabs for this bulk instance.
+        This can be later used to iterate through all slabs,
+        or select one at random, to make a Surface object.
+        """
+        if self.precomputed_slabs_path is not None and self.bulk_id_from_db is not None:
```

Comment:
Alternatively - someone who provides precomputed structures will only be limited to those and MAX_MILLER isn't applied there. I imagine it for the scenario where you're creating it to begin with.

Reply:
> We need to implement a check that what has been saved goes up to the desired miller and generate higher millers if whats stores is too low or filters down to a lower miller otherwise.

Yeah. At the very least, we should probably throw a warning or error if the max_miller passed when running the code is different from that in the precomputed structures even if the intent is to read from precomputed structures. Otherwise this can be a silent source of inconsistencies.

Reply:
Added a check for this: https://github.com/Open-Catalyst-Project/Open-Catalyst-Dataset/blob/refactor-v2/ocdata/core/slab.py#L112